### DISCUSSION TITLE
What is the difference between importing a langium grammar and just putting that grammar into your file?

### DISCUSSION THREAD

**Discussion by Austin-Fulbright** [eclipse-langium/langium#1602] [Q&A] [answered]
So if I have a common file with this inside of it:
```interface Common {
  accDescr?: string;
  accTitle?: string;
  title?: string;
}

fragment TitleAndAccessibilities:
  ((accDescr=ACC_DESCR | accTitle=ACC_TITLE | title=TITLE) EOL)+
;

fragment EOL returns string:
  NEWLINE+ | EOF
;

terminal NEWLINE: /\r?\n/;
terminal ACC_DESCR: /[\t ]*accDescr(?:[\t ]*:([^\n\r]*?(?=%%)|[^\n\r]*)|\s*{([^}]*)})/;
terminal ACC_TITLE: /[\t ]*accTitle[\t ]*:(?:[^\n\r]*?(?=%%)|[^\n\r]*)/;
terminal TITLE: /[\t ]*title(?:[\t ][^\n\r]*?(?=%%)|[\t ][^\n\r]*|)/;

hidden terminal WHITESPACE: /[\t ]+/;
hidden terminal YAML: /---[\t ]*\r?\n(?:[\S\s]*?\r?\n)?---(?:\r?\n|(?!\S))/;
hidden terminal DIRECTIVE: /[\t ]*%%{[\S\s]*?}%%(?:\r?\n|(?!\S))/;
hidden terminal SINGLE_LINE_COMMENT: /[\t ]*%%[^\n\r]*/;

```
and I have a langium file with this:

```grammar GitGraph

import "../common/common";

entry GitGraph:
  NEWLINE*
  'gitGraph' Direction? ':'?
  NEWLINE*
  (
    NEWLINE*
    (TitleAndAccessibilities |
    statements+=Statement |
     NEWLINE)*
  )
;

Statement
: Commit
| Branch
| Merge 
| Checkout 
| CherryPicking
;

Direction:
    dir=('LR' | 'TB' | 'BT');

Commit:
    'commit' 
    (
        'id:' id=STRING 
        |'msg:'? message=STRING
        |'tag:' tags+=STRING
        |'type:' type=('NORMAL' | 'REVERSE' | 'HIGHLIGHT')
    )* EOL;
Branch:
    'branch' name=(ID|STRING)
    ('order:' order=INT)?
    EOL;

Merge:
    'merge' branch=(ID|STRING)
    (
        'id:' id=STRING
        |'tag:' tags+=STRING
        |'type:' type=('NORMAL' | 'REVERSE' | 'HIGHLIGHT')
    )* EOL;

Checkout:
    ('checkout'|'switch') branch=(ID|STRING) EOL;

CherryPicking:
    'cherry-pick' 
    (
        'id:' id=STRING
        |'tag:' tags+=STRING
        |'parent:' parent=STRING
    )* EOL;



terminal INT returns number: /[0-9]+(?=\s)/;
terminal ID returns string: /\w([-\./\w]*[-\w])?/;
terminal STRING: /"[^"]*"|'[^']*'/;

```

what is the difference between just doing this and having the common file directly inside of the grammer gitGraph file. Because for whatever reason it seems to create different rules.

**Comment by msujew** [comment]
Hey @Austin-Fulbright,

the only real difference is that there no guarantee that the order of terminals is preserved. Generally, Terminals are lexed in the order they appear in the grammar. I.e. if a word could be tokenized as two different terminals, the first terminal that appears in the grammar receives precedence.

However, when importing a grammar, all imported terminal rules get put to the back of the list, and consequently get the lowest precedence.

What kind of different behavior are you seeing with the import compared to the inlined grammar?

**Comment by Austin-Fulbright** [comment]
So the grammer rules that involve collecting `TitleAndAccesibilty` tokens. So when I import the `../common/common` it will not recognize any of the title and accesability tokens. For example:
```
gitGraph
accTitle: example acces. title
accDescr: example acces. description
commit
commit
```

inputed into the language with the imports would throw an error such as expecting EOF but found 'accTitle'

However if in the playground or my langium project I use all of the code in the same file, it will work just fine.

There is also some services that I am using from my common. This is probably a misunderstanding on my part how the modules work.

Here are some examples from my project:

[gitGraph/gitGraph.langium](https://github.com/Austin-Fulbright/mermaid/blob/feature/4401_creating_langium_parser_gitGraph/packages/parser/src/language/gitGraph/gitGraph.langium)

[gitGraph/module.ts](https://github.com/Austin-Fulbright/mermaid/blob/feature/4401_creating_langium_parser_gitGraph/packages/parser/src/language/gitGraph/module.ts)

[common/matcher.ts](https://github.com/Austin-Fulbright/mermaid/blob/feature/4401_creating_langium_parser_gitGraph/packages/parser/src/language/common/matcher.ts)

[common/tokenBulder.ts](https://github.com/Austin-Fulbright/mermaid/blob/feature/4401_creating_langium_parser_gitGraph/packages/parser/src/language/common/tokenBuilder.ts)

[common/valueConverter.ts](https://github.com/Austin-Fulbright/mermaid/blob/feature/4401_creating_langium_parser_gitGraph/packages/parser/src/language/common/valueConverter.ts)

Thanks for taking the time to answer my question :)

**Reply by msujew** [reply, marked as answer]
I've put your grammar [into the playground](https://langium.org/playground?grammar=OYJwhgthYgBA4gSwC73ABwBYChuIHbICmIAZmAMZGwDCA9lHfrAN7ayyUUAiRAzhRAB%2BAFyw%2ByEAWABudpwoUAKigA2RUeMnS5HZGo1iJU-LOwBfXKXDAIRQrBXJ1AQXwATF4v59EAI0RVFER%2BEXkACnCuXgEQAF4XGhoAfW4AUQBlGgAlWAAfBWUDBKTkpQBJJQAZNPzYfWciOIrqtIBKWDSAeSq2gGpsOWxrMFt7ZE6e2BAiZABXEHw%2BLRNgMI4AOTSAdSryrb667oAxQdxiEAgCMFVYLd39tLEAegAdYVf8Z7kLq-wb2CJFLpLLZF4AbVeEwAugAqaL8QThUSQmGwkThcEAPU%2B7zhQmRcQApES2nlsbiQHCya8%2BLCWJiseZqeY2t9sL9rrcgWVKjUIVDYHCuE51KihejkSIKfg8bCCUJiaTyTjZVTYWyfiQ-gCWvzYM9xXCGuopUaZXKFUqyebVXK8prcJhEO53PZ6tqubBtgAJSqZAAKiSeBqNfXZztd7s5-1uAE0XABZKovAC06aNsPeQk%2BZteGVp%2BOznzaQnTqalxfweWRAEJ821HZG3cwYwDuOVsmkaBUAGohw2CuEkliQgt8fHmEmVj7VusNpsulsey5ejL7eA1ZJ7LbJGhdROJtIbJQCtEki3q9nYcYgACeCBQaDAWHWdx2O7SsIiAHJgE%2BMEwH86j-ACXyA2AfxEYCClA1BAOgupwjg58sGA7hEBmCh9CYSDoMbeR7k-b8OHCeRNg-R4SI4WBwlFIg3E8bw%2BF8AIgn0fh8nIrQwGIOxCD4Po4gyZBeKIfiJjybj3weLY2motozmwESxIk7AxHoKAUGwAoACFwHwCgcAKRMSGAagdNoTAiAoABrOg5gmSyaGskB7wDRA7OkJSMKwnD8DfDh3EwuJkKqbIYMgpRdMin9dKUH82iGTSrmQQLIIoBhUuA7iyJo-LIJdRCXWEpRsg3WBpJovIfwgPg1h-IRYDsFjRiaDIyo3KqOBq0SGvqUZBNK8qNngbr8h-ZA73QIhEKmmbQp-DYumyRMXCqWKu37bIMjSWK-XgH09kOhLFPy%2BTJiqOR9LAQzMHSn8-AMozgP%2BOxQvKbg8g6kb4DOmjkLoEA3RARCgZBuJ9iUUtuO6K7cFMkBzIeuwkaIYCntuoyPq%2Bn6N3%2B0jxp-IrgJKvHRvG3rRjmwahPJsaCuqybptm4D5qaZDltW9bNrSbbdv28pDuOn1Tu4i64eS6y7IctLcp-IybPsxyfxqvgAHcUBejpMbunHvs60aOkl3AXJIdzPNs6QHsVty71TdBLZy-K8sZwr3GK9xhq6t2Jr6mngCG%2BnKZ-dAYHGRCw5mQhvYp87YUuoZzk9WNYCh6ZZgWJZYHwOYID8EgIQABlTABOaE%2BkJWlHTbW5Poz%2BZFmWYxrYNV51cxVNXgAOjedW4XBLv%2B9Ldla9gemXgAImxSe4UnmrsR-OEf2%2BIA&content=OYSwLg4gTghgDgCwFAwMaoCrgDYFMBcABLgB4wC2cehaquAzgHSFg64roAiDqURpFKrhroGzACY8oIOKwD2AOySo55cuGWr1YIA) and simulated what happens when using grammar imports [here](https://langium.org/playground?grammar=OYJwhgthYgBA4gSwC73ABwBYChuIHbICmIAZmAMZGwDCA9lHfrAN7ayyUUAiRAzhRAB%2BAFyw%2ByEAWABudpwoUAKigA2RUeMnS5HZGo1iJU-LOwBfXKXDAIRQrBXJ1AQXwATF4v59EAI0RVFER%2BEXkACnCuXgEQAF4XGhoAfW4AUQBlGgAlWAAfBWUDBKTkpQBJJQAZNPzYfWciOIrqtIBKWDSAeSq2gGpsOWxrMFt7ZE6e2BAiZABXEHw%2BLRNgMI4AOTSAdSryrb667oAxQdxMRHd3e3qSCAIwVVhtgAlKzIAFRLSxAHoAbQAOhMALp9X5yC5XG7EED3fCPWAATRcAFkqn8ALTYoGggBUgOEgPw4VEQIygL4ILxQkJtPwbSE2MxpJEdOJeVJAEJARk2m0IdgoddmLD4YjuOVsmkaBUAGo-WAA4GwakAUjVLHJlOpQnMGtZ7PwnKEPL5AshlxFtzhDyeGX28BqyT2W2SNC6qNRaQ2Sj%2BuNVeI1-wAesTCdTBdhxiAAJ4IFBoMBYdawLa7fZpPERADkwETGEwObqeYLyaLsBzImLBVLqEL1bq4TrSawxe4iBmFH0TEr1f58nTrqzEXkmx2w%2BzHA44Scrg8XiofF8ASC%2Bn4%2BTHWjAxDshD4fTiGWQO6Ie4meS3aYnmbaU9gbTO2GPp-P2DE9CgKGwBQAQuB8AoHAClREhgGoH9aEwIgKAAazoOYJkgmhoJAOMPkQODpCfDsux7fBUw4dxOziZsqmyGtKyUX9KJzX8lBzNohk-e5kEIysKAYVjiy3cIr2nHNLkbS4jyUbJHVgfiODyHMID4NYcyEWA7GXUYmgyMTHSk-IcxPBT6lGA9RPEjZ4G0mTkFjdAiEbSzrNInMNi6bJURcKpaOlBVsgyNJaLeeAXj2AKGMfacHzxSYqjkf8wEAzB2JzPwAKA4sETsUjym4PINJM%2BBQunZs6BAa4QEbIqSrifYlEZLduii3BQJAcCErsJqiGLJLYqAjKspyx18pnbTBPcYT3GMrSwrCizRlswzDz60zzN0qybOLOymmbJyXLcjy0i8ny-PKAKgpeEKtzvSLmOguCELY3icyAmD4MQnMZL4AB3FAUo6Tq4p67LNNMjo6qukh0Mw2DpASx60NjTF0Ahniwr4yaBKE4sRIWszUek3SZrWubxsWnGdPQGBxkbMmZkIInsenC6QdwXAxTtWAqumWYFiWWB8DmCA-BIf0AAZMQATjBUk4kpC1sBZhEnkyjn5kWZZjChpVAXe8J-kxQEADpfk16kdaNxlBTlxEsb%2BAAiUNrepa2ZNDHNqRzKMLaeIdMz%2BI1zbuVnEhSdIsmyf0VWpaJ%2BEEVkA2pERtbDfAIxpSWNTaPJQ3DEBqXTyk8RYBPzBz8wZY92BA7KSoajD-EuDnIhY7xeOyUT5OhFTtV08zpPs7xUv-flxwq8VZV8QadQY-D7u247rup9b3u8gtIA&content=OYSwLg4gTghgDgCwFAwMaoCrgDYFMBcABLgB4wC2cehaquAzgHSFg64roAiDqURpFKrhroGzACY8oIOKwD2AOySo55cuGWr1YIA). I've mentioned this in my earlier comment:

> However, when importing a grammar, all imported terminal rules get put to the back of the list, and consequently get the lowest precedence.

Since terminals that appear later in the grammar get a lower precedence, the lexer and consequently the parser behave differently. Note that you can adjust the `TokenBuilder` instance to modify that behavior.
